zebra-state/src/service/finalized_state/disk_format/upgrade/add_subtrees.rs

```diff
@@ -0,0 +1,454 @@
+            let mut prev_tree = prev_tree
+                .take()
+                .expect("should have some previous orchard frontier");
+            let orchard_nct = Arc::make_mut(&mut prev_tree);
+
+            let block = upgrade_db
+                .block(height.into())
+                .expect("height with note commitment tree should have block");
+
+            for orchard_note_commitment in block.orchard_note_commitments() {
+                // Return early if there is a cancel signal.
+                if !matches!(cancel_receiver.try_recv(), Err(mpsc::TryRecvError::Empty)) {
+                    return Err(CancelFormatChange);
+                }
+
+                orchard_nct
+                    .append(*orchard_note_commitment)
+                    .expect("finalized notes should append successfully");
+
+                // The loop always breaks on this condition,
+                // because we checked the block has enough commitments,
+                // and that the final commitment in the block doesn't complete a subtree.
+                if orchard_nct.is_complete_subtree() {
+                    break;
+                }
+            }
+
+            let (index, node) = orchard_nct.completed_subtree_index_and_root().expect(
+                "block should have completed a subtree before its final note commitment: \
+                 already checked is_complete_subtree(), and that the block must complete a subtree",
+            );
+
+            assert_eq!(
+                index.0, subtree_count,
+                "trees are inserted in order with no gaps"
+            );
+            write_orchard_subtree(upgrade_db, index, height, node);
+            subtree_count += 1;
+        }
+
+        prev_tree = Some(tree);
+    }
+
+    Ok(())
+}
+
+/// Check that note commitment subtrees were correctly added.
+///
+/// # Panics
+///
+/// If a note commitment subtree is missing or incorrect.
+pub fn check(db: &ZebraDb) {
+    if !check_sapling_subtrees(db) || !check_orchard_subtrees(db) {
+        panic!("missing or bad subtree(s)");
+    }
+}
+
+/// Check that Sapling note commitment subtrees were correctly added.
+///
+/// # Panics
+///
+/// If a note commitment subtree is missing or incorrect.
+fn check_sapling_subtrees(db: &ZebraDb) -> bool {
+    let Some(NoteCommitmentSubtreeIndex(mut first_incomplete_subtree_index)) =
+        db.sapling_tree().subtree_index()
+    else {
+        return true;
+    };
+
+    // If there are no incomplete subtrees in the tree, also expect a subtree for the final index.
+    if db.sapling_tree().is_complete_subtree() {
+        first_incomplete_subtree_index += 1;
+    }
+
+    let mut is_valid = true;
+    for index in 0..first_incomplete_subtree_index {
+        // Check that there's a continuous range of subtrees from index [0, first_incomplete_subtree_index)
+        let Some(subtree) = db.sapling_subtree_by_index(index) else {
+            error!(index, "missing subtree");
```

Comment:
I changed the logs to errors because it's unrecoverable, so we want it big and red.